### PR TITLE
Refactor svmbir loss

### PR DESCRIPTION
**Pull request by smajee**
Fix issue #95 
Fix issue #132

### CODE REVIEW

File: scico/linop/radon_svmbir.py [hunk 1]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
     of :class:`scico.linop.Diagonal`. If :math:`W` is None, it is set to
     :class:`scico.linop.Identity`.
 
+    The extended loss differs from a typical weighted squared :math:`\ell_2` loss
+    is the following ways.
+
     When ``positivity=True``, the prox projects onto the non-negative
     quadrant, but the the loss, :math:`\alpha l(\mb{y}, A(\mb{x}))`,
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:199]: quadrant --> orthant

File: scico/linop/radon_svmbir.py [hunk 2]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
...
     is unaffected by this setting and still evaluates to finite values
     when :math:`\mb{x}` is not in the non-negative quadrant.
+
+    When the ``is_masked`` option of the associated ParallelBeamProjector is True,
+    the reconstruction is computed over a masked region of the image as described
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:204]: Consider adding sphinx `:class:` cross-referencing so that `ParallelBeamProjector` links to the corresponding docs page for that class.

File: scico/linop/radon_svmbir.py [hunk 3]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
...
+    in the ParallelBeamProjector class.
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:205]: As above, possibly with rephrasing, e.g. in class :class:ParallelBeamProjector (backticks omitted due to conflict with markup here)

File: scico/linop/radon_svmbir.py [hunk 4]
```diff
@@ -212,13 +222,10 @@ def __init__(
...
-                :meth:`svmbir.recon` prox routine. Note that omitting
-                this dictionary will cause the default dictionary to be
-                used, however omitting entries in the passed dictionary
-                causes the defaults of the underlying :meth:`svmbir.recon`
-                prox routine to be used.
+                :meth:`svmbir.recon` prox routine. Defaults to
+                {"maxiter": 1000, "ctol": 0.001}.
             positivity: Enforce positivity in the prox operation. The
-                loss is not affected.
+                loss is infinity if any element of the input is negative.
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:228]: Perhaps "loss is infinite"?

File: scico/linop/radon_svmbir.py [hunk 5]
```diff
@@ -230,15 +237,35 @@ def __init__(
         if prox_kwargs is None:
             prox_kwargs = dict()
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:238]: Egregious nitpick: I think ```prox_kwargs = {}``` would be a more standard way of writing this.

File: scico/linop/radon_svmbir.py [hunk 6]
```diff
@@ -230,15 +237,35 @@ def __init__(
...
+
         svmbir_prox_args = dict()
-        if "maxiter" in prox_kwargs:
-            svmbir_prox_args["max_iterations"] = prox_kwargs["maxiter"]
-        if "ctol" in prox_kwargs:
-            svmbir_prox_args["stop_threshold"] = prox_kwargs["ctol"]
+        if "maxiter" in default_prox_args:
+            svmbir_prox_args["max_iterations"] = default_prox_args["maxiter"]
+        if "ctol" in default_prox_args:
+            svmbir_prox_args["stop_threshold"] = default_prox_args["ctol"]
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:247]: This approach makes sense for just two options to be passed to svmbir, but if that were ever to expand, I would suggest switching to a loop over a dict mapping our parameter names to svmbir ones.

File: scico/linop/radon_svmbir.py [hunk 7]
```diff
@@ -230,15 +237,35 @@ def __init__(
...
         self.positivity = positivity
 
+        if W is None:
+            self.W = Identity(self.y.shape)
+        elif isinstance(W, Diagonal):
+            if snp.all(W.diagonal >= 0):
+                self.W = W
+            else:
+                raise Exception(f"The weights, W.diagonal, must be non-negative.")
+        else:
```
Comment (bwohlberg) [inline scico/linop/radon_svmbir.py:259]: Perhaps just "The weights, W, must be ..." since W.diagonal is internal and not part of the interface?

**Review by bwohlberg — APPROVED** [review]
Looks good to me, but it would be best to get at least one other review.

File: scico/linop/radon_svmbir.py (hunk 6 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:247]: Thanks. I will keep that in mind as we change the svmbir interface.

File: scico/linop/radon_svmbir.py (hunk 7 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:256]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 5 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 4 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 1 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 2 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:204]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 3 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed.

File: scico/linop/radon_svmbir.py [hunk 8]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
     of :class:`scico.linop.Diagonal`. If :math:`W` is None, it is set to
     :class:`scico.linop.Identity`.
 
+    The extended loss differs from a typical weighted squared :math:`\ell_2` loss
+    is the following ways.
```
Comment (tbalke) [inline scico/linop/radon_svmbir.py:196]: in the following ways?

File: scico/linop/radon_svmbir.py (hunk 1 above)
Comment (tbalke) [inline scico/linop/radon_svmbir.py]: the the

File: scico/linop/radon_svmbir.py [hunk 9]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
...
+
     When ``positivity=True``, the prox projects onto the non-negative
     quadrant, but the the loss, :math:`\alpha l(\mb{y}, A(\mb{x}))`,
     is unaffected by this setting and still evaluates to finite values
     when :math:`\mb{x}` is not in the non-negative quadrant.
+
+    When the ``is_masked`` option of the associated ParallelBeamProjector is True,
```
Comment (tbalke) [inline scico/linop/radon_svmbir.py]: `True`

File: scico/linop/radon_svmbir.py (hunk 3 above)
Comment (tbalke) [inline scico/linop/radon_svmbir.py]: :class:`ParallelBeamProjector`

**Review by tbalke — APPROVED** [review]
Approve upon the few minor corrections

File: scico/linop/radon_svmbir.py [hunk 10]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
...
+    When the ``is_masked`` option of the associated :class:`.ParallelBeamProjector` is True,
+    the reconstruction is computed over a masked region of the image as described
+    in class :class:`.ParallelBeamProjector`.
     """
 
     def __init__(
         self,
         *args,
-        prox_kwargs: dict = {"maxiter": 1000, "ctol": 0.001},
+        prox_kwargs: dict = {},
```
Comment (Michael-T-McCann) [inline scico/linop/radon_svmbir.py:211]: initializing with a mutable type is risky, I suggest `=None` then `if prox_kwargs is None:`.

File: scico/linop/radon_svmbir.py [hunk 11]
```diff
@@ -190,20 +192,28 @@ class SVMBIRWeightedSquaredL2Loss(WeightedSquaredL2Loss):
...
 
+    The extended loss differs from a typical weighted squared :math:`\ell_2` loss
+    is the following ways.
+
     When ``positivity=True``, the prox projects onto the non-negative
-    quadrant, but the the loss, :math:`\alpha l(\mb{y}, A(\mb{x}))`,
+    orthant, but the the loss, :math:`\alpha l(\mb{y}, A(\mb{x}))`,
     is unaffected by this setting and still evaluates to finite values
-    when :math:`\mb{x}` is not in the non-negative quadrant.
+    when :math:`\mb{x}` is not in the non-negative orthant.
```
Comment (Michael-T-McCann) [inline scico/linop/radon_svmbir.py:201]: "nonnegative". Generally don't hyphenate prefixes (unless they are words themselves, or if needed for clarity, or...)

File: scico/linop/radon_svmbir.py [hunk 12]
```diff
@@ -212,13 +222,10 @@ def __init__(
...
-                :meth:`svmbir.recon` prox routine. Note that omitting
-                this dictionary will cause the default dictionary to be
-                used, however omitting entries in the passed dictionary
-                causes the defaults of the underlying :meth:`svmbir.recon`
-                prox routine to be used.
+                :meth:`svmbir.recon` prox routine. Defaults to
+                {"maxiter": 1000, "ctol": 0.001}.
             positivity: Enforce positivity in the prox operation. The
-                loss is not affected.
+                loss is infinite if any element of the input is negative.
```
Comment (Michael-T-McCann) [inline scico/linop/radon_svmbir.py:228]: contradicts earlier statement of "unaffected"?

**Comment by tbalke** [comment]
Perhaps we also want to add an example for the SVMBIRExtendedLoss?

File: scico/linop/radon_svmbir.py [hunk 13]
```diff
@@ -227,18 +234,35 @@ def __init__(
...
 
         self.positivity = positivity
 
+        if W is None:
+            self.W = Identity(self.y.shape)
+        elif isinstance(W, Diagonal):
+            if snp.all(W.diagonal >= 0):
+                self.W = W
+            else:
+                raise Exception(f"The weights, W, must be non-negative.")
```
Comment (Michael-T-McCann) [inline scico/linop/radon_svmbir.py:255]: This is checked by the `__init__` of `WeightedSquaredL2Loss`. Would it be possible to reduce duplication by making use of `super().__init__`?

**Review by Michael-T-McCann — APPROVED** [review]
Approved pending a couple minor comments.

File: scico/linop/radon_svmbir.py (hunk 8 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:196]: How about "... in the following aspects"?

File: scico/linop/radon_svmbir.py (hunk 10 above)
Comment (tbalke) [reply scico/linop/radon_svmbir.py:211]: Check out `admm.py`

```python
    def __init__(self, cg_kwargs: Optional[dict] = None, cg_function: str = "scico"):
        """Initialize a :class:`LinearSubproblemSolver` object.

        Args:
            cg_kwargs: Dictionary of arguments for CG solver. See
                documentation for :func:`scico.solver.cg` or
                :func:`jax.scipy.sparse.linalg.cg`,
                including how to specify a preconditioner.
                Default values are the same as those of
                :func:`scico.solver.cg`, except for
                ``"tol": 1e-4`` and ``"maxiter": 100``.
            cg_function: String indicating which CG implementation to
                use. One of "jax" or "scico"; default "scico".  If
                "scico", uses :func:`scico.solver.cg`. If "jax", uses
                :func:`jax.scipy.sparse.linalg.cg`. The "jax" option is
                slower on small-scale problems or problems involving
                external functions, but can be differentiated through.
                The "scico" option is faster on small-scale problems, but
                slower on large-scale problems where the forward
                operator is written entirely in jax.
        """

        default_cg_kwargs = {"tol": 1e-4, "maxiter": 100}
        if cg_kwargs:
            default_cg_kwargs.update(cg_kwargs)
        self.cg_kwargs = default_cg_kwargs
```

File: scico/linop/radon_svmbir.py (hunk 8 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Changed it.

File: scico/linop/radon_svmbir.py (hunk 1 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed.

File: scico/linop/radon_svmbir.py (hunk 9 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed

File: scico/linop/radon_svmbir.py (hunk 3 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed

File: scico/linop/radon_svmbir.py (hunk 11 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed

File: scico/linop/radon_svmbir.py (hunk 12 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:224]: Good catch! Fixed it now.

File: scico/linop/radon_svmbir.py (hunk 13 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py:251]: The `SVMBIRExtendedLoss` inherits from `Loss` instead of `WeightedSquaredL2Loss` since it is not a true WeightedSquaredL2Loss. So `super().__init__` won't work here since `Loss` doesn't have the weight.

One Way to deduce duplication would be to define a `WeightedLoss ` that inherits from loss. Both `WeightedSquaredL2Loss` and `SVMBIRExtendedLoss` can inherit from this.

File: scico/linop/radon_svmbir.py (hunk 10 above)
Comment (smajee) [reply scico/linop/radon_svmbir.py]: Fixed it.

**Comment by smajee** [comment]
> Perhaps we also want to add an example for the SVMBIRExtendedLoss?

Talked with @bwohlberg  about it. He seemed to think that it wasn't necessary since there would be a lot of code duplication between the `SVMBIRExtendedLoss` and `SVMBIRWeightedSquaredL2Loss` examples.
